Update event-driven-content.md (#4513)
MicrosoftDocs/architecture-center: docs/guide/architecture-styles/event-driven-content.md

In the logical diagram above, each type of consumer is shown as a single box. In practice, it's common to have multiple instances of a consumer, to avoid having the consumer become a single point of failure in system. Multiple instances might also be necessary to handle the volume and frequency of events. Also, a single consumer might process events on multiple threads. This can create challenges if events must be processed in order or require exactly-once semantics. See [Minimize Coordination][minimize-coordination].

There are two primary topologies within the event-driven architecture:

> - **Broker topology**. Components broadcast occurrences as events to the entire system, and other components either act upon the event or just ignore the event. This topology is useful when the workflow is relatively simple. There is no central coordination or orchestration so this topology is very dynamic based on various conditions. This topology is highly decoupled, provides high scalability, high responsiveness, high performance and high fault tolerance. On the other hand, no component owns or is aware of the state of the business transaction and actions are taken asynchronously, subsequently no component can restart the business transaction. In this topology, error handling is a challenge and may require manual intervention. Also this topology is more likely to cause data inconsistency.

Author: Clarified it in a new commit

> - **Mediator topology**. This topology addresses some of the shortcomings of broker topology. There is an event mediator that manages and controls the workflow of events. The event mediator maintains the state, manages error handling and restart capabilities. Unlike broker topology, components broadcast occurrences as commands and only to designated channels (usually message queues) which can’t be ignored; the component on the receiving end must act upon the command. This topology offers better control, better error handling and better data consistency. On the other hand, this topology has more complings between components, offers lower scalability (because the event mediator can be a bottleneck), lower performance and lower fault tolerance. Also, it can be difficult to model complex workflows. 

Reviewer: ```suggestion
- **Mediator topology**. This topology addresses some of the shortcomings of broker topology. There is an event mediator that manages and controls the workflow of events. The event mediator maintains the state, manages error handling and restart capabilities. Unlike broker topology, components broadcast occurrences as commands and only to designated channels, usually message queues. These commands are not expected to be ignored by their consumers. This topology offers more control, better distributed error handling, and potentially better data consistency. This topology does introduce increased coupling between components and the event mediator could become a bottleneck or a reliability concern.
```

## When to use this architecture

- Multiple subsystems must process the same events.
- Real-time processing with minimum time lag.
- Complex event processing, such as pattern matching or aggregation over time windows.
- High volume and high velocity of data, such as IoT.

> If you need workflow control and error handling capability, you should use mediator topology. But if you need higher performance and scalability, you should choose broker topology. You can also use a hybrid model combining both mediator and broker topologies.
> 

Reviewer: Let's leave this part out for now.  It really doesn't fit under "When to use this architecture" anyhow.

```suggestion
```

Author: Personally, I think this paragraph makes a nice summary of differences and also informs readers about the possibility of hybrid approach. To accommodate your opinion, I've moved this paragraph to upper section. But I leave it up to you whether to keep or discard it.

In the logical diagram above, each type of consumer is shown as a single box. In practice, it's common to have multiple instances of a consumer, to avoid having the consumer become a single point of failure in system. Multiple instances might also be necessary to handle the volume and frequency of events. Also, a single consumer might process events on multiple threads. This can create challenges if events must be processed in order or require exactly-once semantics. See [Minimize Coordination][minimize-coordination].

> There are two primary topologies within the event-driven architecture:

Reviewer: ```suggestion
There are two primary topologies within many event-driven architectures:
```

> - **Broker topology**. Components broadcast occurrences as events to the entire system, and other components either act upon the event or just ignore the event. This topology is useful when event processing flow is relatively simple. There is no central coordination or orchestration, so this topology can be very dynamic. This topology is highly decoupled which helps provide scalability, responsiveness, and component fault tolerance. No component owns or is aware of the state of any multi-step business transaction and actions are taken asynchronously. Subsequently, distributed transactions are risky because there is no native means to be restarted or replayed. Error handling and manual intervention strategies needs to be carefully considered as this topology can be a source of data inconsistency.

Reviewer 2: ```suggestion
- **Broker topology**. Components broadcast occurrences as events to the entire system, and other components either act upon the event or just ignore the event. This topology is useful when the event processing flow is relatively simple. There is no central coordination or orchestration, so this topology can be very dynamic. This topology is highly decoupled, which helps provide scalability, responsiveness, and component fault tolerance. No component owns or is aware of the state of any multistep business transaction, and actions are taken asynchronously. Subsequently, distributed transactions are risky because there is no native means to be restarted or replayed. Error handling and manual intervention strategies need to be carefully considered because this topology can be a source of data inconsistency.
```

> - **Mediator topology**. This topology addresses some of the shortcomings of broker topology. There is an event mediator that manages and controls the flow of events. The event mediator maintains the state, manages error handling and restart capabilities. Unlike broker topology, components broadcast occurrences as commands and only to designated channels, usually message queues. These commands are not expected to be ignored by their consumers. This topology offers more control, better distributed error handling, and potentially better data consistency. This topology does introduce increased coupling between components and the event mediator could become a bottleneck or a reliability concern.

Reviewer 2: ```suggestion
- **Mediator topology**. This topology addresses some of the shortcomings of broker topology. There is an event mediator that manages and controls the flow of events. The event mediator maintains the state and manages error handling and restart capabilities. Unlike broker topology, components broadcast occurrences as commands and only to designated channels, usually message queues. These commands aren't expected to be ignored by their consumers. This topology offers more control, better distributed error handling, and potentially better data consistency. This topology does introduce increased coupling between components, and the event mediator could become a bottleneck or a reliability concern.
```

- **Mediator topology**. This topology addresses some of the shortcomings of broker topology. There is an event mediator that manages and controls the flow of events. The event mediator maintains the state, manages error handling and restart capabilities. Unlike broker topology, components broadcast occurrences as commands and only to designated channels, usually message queues. These commands are not expected to be ignored by their consumers. This topology offers more control, better distributed error handling, and potentially better data consistency. This topology does introduce increased coupling between components and the event mediator could become a bottleneck or a reliability concern.

> If you need error handling capability and control over flow of events, you should use mediator topology. But if you need higher performance and higher scalability, you should choose broker topology. You can also use a hybrid model combining both mediator and broker topologies.
> 

Reviewer: We haven't even told them "when to use this architecture" yet -- vs even deciding on which approach to use.  As I suggested before, lets leave this out.  Especially since the new topology text doesn't even have "performance" in there yet it's here in the "choose" text.

```suggestion
```
